### PR TITLE
Support alternate key 2.0

### PR DESCRIPTION
**Pull request by xuzhg**
Merged #219, Resolve the comments and add more test cases.

### CODE REVIEW

**Comment by xuzhg** [comment]
Merged the PR. Thanks.

File: src/Microsoft.OData.Core/UriParser/Binders/KeyBinder.cs [hunk 1]
```diff
@@ -41,21 +41,101 @@ internal KeyBinder(MetadataBinder.QueryTokenVisitor keyValueBindMethod)
...
-            List<KeyPropertyValue> keyPropertyValues = new List<KeyPropertyValue>();
 
             IEdmEntityType collectionItemEntityType = collectionItemType.EntityDefinition();
+            QueryNode keyLookupNode;
 
+            if (TryBindToDeclaredKey(collectionNode, namedValues, model, collectionItemEntityType, out keyLookupNode))
+            {
+                return keyLookupNode;
+            }
+            else if (TryBindToDeclaredAlternateKey(collectionNode, namedValues, model, collectionItemEntityType, out keyLookupNode))
```
Comment (congysu) [inline src/Microsoft.OData.Core/UriParser/Binders/KeyBinder.cs]: The bind key logic does not seem to be correct.

File: src/Microsoft.OData.Core/UriParser/Metadata/AlternateKeysODataUriResolver.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        /// </summary>
+        /// <param name="type">Type for current entityset.</param>
+        /// <param name="namedValues">The dictionary of name value pairs.</param>
+        /// <param name="keyProperties">Dictionary of alias to key properties.</param>
+        /// <param name="convertFunc">The convert function to be used for value converting.</param>
+        /// <param name="convertedPairs">The resolved key list.</param>
+        /// <returns>True if resolve succeeded.</returns>
+        private bool TryResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, IDictionary<string, IEdmProperty> keyProperties, Func<IEdmTypeReference, string, object> convertFunc, out IEnumerable<KeyValuePair<string, object>> convertedPairs)
+        {
+            Dictionary<string, object> pairs = new Dictionary<string, object>(StringComparer.Ordinal);
```
Comment (congysu) [inline src/Microsoft.OData.Core/UriParser/Metadata/AlternateKeysODataUriResolver.cs]: var

File: src/Microsoft.OData.Core/UriParser/Metadata/AlternateKeysODataUriResolver.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                if (EnableCaseInsensitive)
+                {
+                    var list = namedValues.Keys.Where(key => string.Equals(kvp.Key, key, StringComparison.OrdinalIgnoreCase)).ToList();
+                    if (list.Count > 1)
+                    {
+                        throw new ODataException(Strings.UriParserMetadata_MultipleMatchingKeysFound(kvp.Key));
+                    }
+                    else if (list.Count == 0)
+                    {
+                        convertedPairs = null;
```
Comment (congysu) [inline src/Microsoft.OData.Core/UriParser/Metadata/AlternateKeysODataUriResolver.cs]: Move `convertedPairs = null;` to the top of the method.

File: src/Microsoft.OData.Edm/Csdl/EdmxReader.cs [hunk 4]
```diff
@@ -357,9 +358,9 @@ private List<CsdlModel> LoadAndParseReferencedEdmxFiles(Version mainEdmxVersion)
                     continue;
                 }
 
-                if (tmp.Uri != null
-                    && (tmp.Uri.EndsWith(CoreVocabularyConstants.VocabularyUrlSuffix, StringComparison.Ordinal))
-                        || tmp.Uri.EndsWith(CapabilitiesVocabularyConstants.VocabularyUrlSuffix, StringComparison.Ordinal))
+                if (tmp.Uri != null && (tmp.Uri.EndsWith(CoreVocabularyConstants.VocabularyUrlSuffix, StringComparison.Ordinal) ||
+                    tmp.Uri.EndsWith(CapabilitiesVocabularyConstants.VocabularyUrlSuffix, StringComparison.Ordinal) ||
+                    tmp.Uri.EndsWith(CommunityVocabularyConstants.VocabularyUrlSuffix, StringComparison.Ordinal)))
```
Comment (congysu) [inline src/Microsoft.OData.Edm/Csdl/EdmxReader.cs]: Will we need to add another `||`, if there is another community vocabulary, such as Display?  We shall load, and optionally cache a community vocabulary.

File: src/Microsoft.OData.Edm/Library/EdmModelBase.cs [hunk 5]
```diff
@@ -48,6 +49,11 @@ protected EdmModelBase(IEnumerable<IEdmModel> referencedModels, IEdmDirectValueA
                 this.referencedEdmModels.Add(CapabilitiesVocabularyModel.Instance);
             }
 
+            if (CommunityVocabularyModel.Instance != null)
```
Comment (congysu) [inline src/Microsoft.OData.Edm/Library/EdmModelBase.cs]: We need to add another `if` clause, if there is a new community vocabulary.  This does not seem to be right.

**Pull request closed without merging**